elements/auras.lua

```diff
@@ -422,17 +508,96 @@ local function UpdateAuras(self, event, unit)
 	if(buffs) then
 		if(buffs.PreUpdate) then buffs:PreUpdate(unit) end
 
+		local buffsChanged = false
 		local numBuffs = buffs.num or 32
-		local visibleBuffs = filterIcons(buffs, unit, buffs.filter or 'HELPFUL', numBuffs)
-		buffs.visibleBuffs = visibleBuffs
+		local buffFilter = buffs.filter or 'HELPFUL'
+		if(type(buffFilter) == 'function') then
+			buffFilter = buffFilter(buffs, unit)
+		end
+
+		if(not updateInfo or updateInfo.isFullUpdate) then
+			buffs.active = table.wipe(buffs.active or {})
+			buffs.sorted = table.wipe(buffs.sorted or {})
+			buffsChanged = true
+
+			local slots = {UnitAuraSlots(unit, buffFilter)}
+			if(slots[2]) then -- #1 return is continuationToken, we don't care about it
+				local count = 1
+
+				for i = 2, #slots do
+					if count <= numBuffs then
+						local data = processData(C_UnitAuras.GetAuraDataBySlot(unit, slots[i]))
+						if((buffs.FilterAura or FilterAura) (buffs, unit, data)) then
+							buffs.active[data.auraInstanceID] = data
+
+							table.insert(buffs.sorted, data)
+
+							count = count + 1
+						end
+					end
+				end
+			end
+		else
+			if(updateInfo.updatedAuraInstanceIDs) then
+				for _, auraInstanceID in next, updateInfo.updatedAuraInstanceIDs do
+					if(buffs.active[auraInstanceID]) then
+						buffs.active[auraInstanceID] = processData(C_UnitAuras.GetAuraDataByAuraInstanceID(unit, auraInstanceID))
+						buffsChanged = true
+					end
+				end
+			end
+
+			local buffCount = #buffs.sorted
+
+			if(updateInfo.removedAuraInstanceIDs) then
+				for _, auraInstanceID in next, updateInfo.removedAuraInstanceIDs do
+					if(buffs.active[auraInstanceID]) then
+						buffs.active[auraInstanceID] = nil
+						buffCount = buffCount - 1
+						buffsChanged = true
+					end
+				end
+			end
+
+			if(updateInfo.addedAuras) then
+				for _, data in next, updateInfo.addedAuras do
+					if(data.isHelpful and not C_UnitAuras.IsAuraFilteredOutByInstanceID(unit, data.auraInstanceID, buffFilter)) then
+						-- only try to add new buffs if we have enough room for them
+						if(buffCount <= numBuffs) then
+							data = processData(data)
+							if((buffs.FilterAura or FilterAura) (buffs, unit, data)) then
+								buffs.active[data.auraInstanceID] = data
+								buffCount = buffCount + 1
+								buffsChanged = true
+							end
+						end
+					end
+				end
+			end
+
+			if(buffsChanged) then
+				buffs.sorted = table.wipe(buffs.sorted or {})
+
+				for _, data in next, buffs.active do
+					table.insert(buffs.sorted, data)
+				end
+
+				table.sort(buffs.sorted, buffs.SortBuffs or buffs.SortAuras or SortAuras)
+			end
+		end
 
-		local fromRange, toRange
-		if(buffs.PreSetPosition) then
-			fromRange, toRange = buffs:PreSetPosition(numBuffs)
+		if(not buffsChanged) then return end
```

Comment:
I believe this is causing `UpdateAuras` to exit early if there have been no buff changes even if there have been debuff changes, resulting in debuffs not being added or removed.

Reply:
Ehm, it's in `.Buffs`, so obv it only handles buffs...

Reply:
I could totally be misunderstanding how it's supposed to be used, but it appears that if a layout has set both `.Buffs` and `.Debuffs`, it relies on `UpdateAuras` to update both of them, not just the buffs. In this case, we are returning from `UpdateAuras` if there are no changes to buffs even if there are changes to debuffs.

Reply:
Ah, true >_> Good catch. More work 😒🔫